added get-started-experiments generation code

iesahin opened the pull request:
This PR contains the source code and generation script for https://github.com/iterative/get-started-experiments

* This is split of #35 
* Related with #34 

It includes the general `generate.bash` script that prepares the `build` directory, calls the `generate-experiments.bash` and writes a _push script_ that pushes the tags to the repository. 

- [x] merge `generate.bash` and `generate-experiments.bash` in a single file and move the directory one level up
- [x] push the resulting `models/fashion-mnist/model.h5` to remote
- [x] check and update the remote per #48 
- [x] prepared and preprocessed data should not be downloaded from remote (#54)
- [x] merge `prepare` and `preprocess` stages to reduce the intermediate file size and use the cache ([Comment](https://github.com/iterative/example-repos-dev/issues/54#issuecomment-855731329) in #54)
- [x] remove `fashion-mnist` intermediate directory from `data` and `models` to reduce the final depth
- [x] rename the directory to `get-started-experiments`
- [x] update the Readme file
- [x] simplify initial `.gitignore` 
- [x] move `venv` instructions in a `<detail>` block
- [x] modify according to this [review](https://github.com/iterative/example-repos-dev/pull/44#pullrequestreview-673582057)

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+and limit the parameters and metrics to show with `--include-params` and
+`--include-metrics` options, respectively.  
+
+Experiments are given hash-names derived from their inputs and environment. It
```
"Hash-names?" Is this this official DVC term for this? CC @jorgeorpinel.

Also "derived from their inputs and environment" - can we be more specific or link to documentation explaining the naming convention?

casperdcl replied:
https://dvc.org/doc/command-reference/exp/run#options are you looking for `dvc exp run --name <name>`?

pmrowla replied:
The autogenerated names are based on a SHA256 hash of the reproduced pipeline. They are hashed in the same way that we generate hashes for stages which are used in run-cache, except that we hash the entire pipeline instead of individual stages.

The hashes really only have meaning internally to DVC, I'm not sure it needs to be documented anywhere outside of the DVC code base itself. If users want experiment names to mean something they should be providing their own names via `-n/--name`.

jorgeorpinel replied:
> Is this this official DVC term for this?

We just use "names" now. See https://dvc.org/doc/command-reference/exp/run. I see no need to go into any details on how they're named here TBH.

```suggestion
Experiments are given automatic names but it
```

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+`--include-metrics` options, respectively.  
+
+Experiments are given hash-names derived from their inputs and environment. It
+may be easier to review them when you give names with `--name/-n` option.
```
```suggestion
may be easier to review them when you give names with the `--name/-n` option.
```

casperdcl commented on get-started/code-experiments/README.md:
````diff
@@ -0,0 +1,260 @@
+dvc exp run -n baseline-experiment
+```
+
+Experiments are normally not checked-in to Git. When there is an experiment that
+you want to preserve in Git history, you can use:
````
```suggestion
Experiments are normally not preserved by Git. If you want to do so, you can use:
```

casperdcl commented on get-started/code-experiments/README.md:
````diff
@@ -0,0 +1,260 @@
+dvc exp apply exp-123456
+```
+
+Then you can use `git add`, `git commit`, `git push` and `dvc push` as usual. 
````
```suggestion
followed by `git add`, `git commit`, `git push` and `dvc push` as usual. 
```

iesahin replied:
I don't want to break the sentence by a command.

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+
+Then you can use `git add`, `git commit`, `git push` and `dvc push` as usual. 
+
+You can push and pull the _code changes_ related to an experiment with `dvc exp
+push` and `dvc exp pull` respectively. These two commands work with _Git
+remotes._
```
```suggestion
You can push and pull the _code changes_ related to an experiment with
`dvc exp push` and `dvc exp pull` respectively. These two commands work with
_Git remotes._
```

iesahin replied:
`p` of `dvc exp` is on column 79, why do you break? MD converter should be smart enough to convert `dvc exp
push`

jorgeorpinel replied:
Yeah I guess it doesn't matter here (README file). It would need to be broken if it was content for the docs engine though (per https://dvc.org/doc/user-guide/contributing/docs).

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+push` and `dvc exp pull` respectively. These two commands work with _Git
+remotes._
+
+You can clean up the experiments that don't make into the repository with:
```
```suggestion
You can clean up unused experiments with:
```

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+## DVC Get Started for Experiments
```
I'm sure @jorgeorpinel will have strong opinions on titles & section titles to remain consistent with the current docs.

iesahin replied:
_The DVC Get Started Project Focusing Mainly on Experiments (but may be used for other features too)_. 

We have two pages for the experiments in the docs, one is titled "Get Started: Experiments", and the other is "Experiment Management." "Get Started: Experiments" looks like a document title instead of a project name but I can use it here too @jorgeorpinel

jorgeorpinel replied:
I think @casperdcl meant to keep just **DVC Get Started** here, as it is now (not sure).

Idk, is this repo now only intended to be used in the GS/Experiments and UG/Exp Mgmt pages? If so maybe **DVC Experiments Example** would be a better title.

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+### Pipeline and Parameters
+
+The pipeline defined in `dvc.yaml` consists of four stages that depend on each
+other sequentially: `prepare`, `preprocess`, `train` and `evaluate`. `train`
```
```suggestion
other sequentially: `prepare`, `preprocess`, `train` and `evaluate`. The `train`
```

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+
+The pipeline defined in `dvc.yaml` consists of four stages that depend on each
+other sequentially: `prepare`, `preprocess`, `train` and `evaluate`. `train`
+stage also depends on `src/models.py` that defines two simple models, `mlp` and
+`cnn`.  
```
```suggestion
stage also depends on `src/models.py` (which defines an MLP and CNN model).
```

iesahin replied:
Again, `mlp` and `cnn` are names here. It's like naming your dragon as `dragon` but it looks easier this way.

jorgeorpinel replied:
Yes but again the syntax is better in the suggestion — "which" clause in parenthesis (could be a comma too).

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+stage also depends on `src/models.py` that defines two simple models, `mlp` and
+`cnn`.  
+
+When you change a parameter for an earlier stage, later stages depend on this
+stage are also run by `dvc exp run`. So when you change a parameter for
+`prepare`, all the later stages are run. 
```
```suggestion
When you change a stage's parameters, all dependent stages are also run by
`dvc exp run`. For example, changing a parameter for `prepare` will result in
all the later stages being run.
```

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+- `prepare.remix`: Determines whether Fashion-MNIST train (60000 images) and
+  test (10000 images) sets are merged and split. If `false`, the test and
+  train sets are not merged and used as in the original.
+
```
inter-bullet spacing is annoying in this long list - please remove

iesahin replied:
OK.I would like to use definition lists but this is not RST.

jorgeorpinel replied:
I agree since most elements are short sentences the bullet list doesn't need spacing but "is annoying" wasn't a good reason haha. Annoying to whom? Users see rendered markdown 🙂

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+
+Notable parameters for each stage that you can use with `--set-param` option are
+as follows: 
+
```
```suggestion

<details>
<summary>Parameters</summary>

```

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+- `tp`: Number of True Positives
+
+- `tn`: Number of True Negatives
+
```
```suggestion

</details>

```

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+configured remote.
+
+- `data/fashion-mnist/raw.dvc`: Contains a reference to the [Dataset
+  Registry][dsr] to download the Fashion-MNIST
```
missing/broken link

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+
+### Source Files
+
+The source files are `src/` directory. All files receive runtime parameters from
```
```suggestion
The source files are in the `src/` directory. All files receive runtime parameters from
```

casperdcl commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+
+The source files are `src/` directory. All files receive runtime parameters from
+`params.yaml`, so none of them require any options. File dependencies are
+hardcoded in the current version, but this may change in a later iteration.
```
this sounds like a note to self rather than info for users?

iesahin replied:
Removed

casperdcl reviewed (changes requested):
Initial review (mostly just `*.md`) for now.

- Since you use `vim`, can you use a plugin which strips trailing spaces (line ending) as well as end-of-file (single-newline).
- Also don't use `[links][like this]` unless you re-use them. Otherwise it's just a liability (broken/missing links) - spotted at least one I think.

jorgeorpinel commented:
May want @dberenbaum's review instead of mine in this one 🙂 

Just one minor Q: is "generate.bash" Bash-specific or is it a general Shell script (.sh extension) ?

jorgeorpinel commented on get-started/generate-experiments.bash:
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
```
```suggestion
#!/usr/bin/env bash
```

But will it work if I use Zsh? Probably in general, just wondering if there's a more general shebang.

Very miinor

iesahin replied:
My experience with Termux on Android led me to believe that finding `/usr/bin/env` in that exact location has the same probability as `/bin/bash`, so I started skipping this but you're right, that's a better practice. 

Fixed. 

Anyone who uses Zsh probably has Bash also installed. Bash is almost universal now and this particular script should run on both. The shell runs `/bin/bash generate.bash` (or `/usr/bin/env bash generate.bash`) when I call `./generate.bash`, so it doesn't matter if the user calls this from Zsh or Fish or any other shell.

aguschin commented on get-started/code-experiments/README.md:
````diff
@@ -0,0 +1,260 @@
+python -m venv .venv
+. .venv/bin/activate
+python -m pip install -r requirements.txt
+```
````
We should add `dvc pull` here, otherwise next command (`dvc exp run...`) fails.

iesahin replied:
Thank you @aguschin. 👍🏼  Fixed now.

iesahin commented on get-started/code-experiments/.gitignore:
```diff
@@ -0,0 +1,51 @@
+# Byte-compiled / optimized / DLL files
```
I'm using Github's default Python `.gitignore`. I'd have this way to keep for users to have different `venv` dirs or whatnot.

casperdcl replied:
For an example repo I think think this is unneeded bloat distracting from the point (should be "how to config DVC (ignoring ignorable files explicitly created in this example)" not "how to config a project (with build config and pre-commit hooks and ignoring ignorable editor config)") - probably best to hear from others and go with majority view.

jorgeorpinel replied:
It's a stark contrast to the current one indeed: https://github.com/iterative/example-get-started/blob/master/.gitignore

iesahin commented:
>     * Since you use `vim`, can you use a plugin which strips trailing spaces (line ending) as well as end-of-file (single-newline)?

Actually I used VS Code with Remote editing in this PR. Vim has these config, I'll take care of this.

iesahin commented:
>     * Also don't use `[links][like this]` unless you re-use them. Otherwise it's just a liability (broken/missing links) - spotted at least one I think.

I'm using these because of readability. I read these documents usually as text. They fit into the paragraph and don't distract the flow. Broken links are a different matter, though, I don't put the links while writing and forget to put them sometimes. They are easy to spot though.

iesahin commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+
+[fmnist]: https://github.com/zalandoresearch/fashion-mnist
+
+It is used as a showcase for [`dvc exp`][gsdvcexp] commands to manage large
+number of experiments. 
+
+[gsdvcexp]: https://dvc.org/doc/start/experiments
```
"large _number_ of experiments" or "large _numbers_ of experiments"? The experiments, in my mind, consist of one group with large number of experiments, not multiple groups having large number of experiments.

jorgeorpinel replied:
Either "a large number" or "large numbers" are correct I think.

iesahin commented on get-started/code-experiments/README.md:
````diff
@@ -0,0 +1,260 @@
+[installdvc]: https://dvc.org/doc/install
+
+```console
+python -m venv .venv
+. .venv/bin/activate
````
[Previous example-get-started](https://github.com/iterative/example-get-started) begins literally from `git clone` and `cd`. 😆 

We even have [an issue](https://github.com/iterative/example-repos-dev/issues/26) regarding to use `python -m venv` instead of `virtualenv` 😄 

I think when someone knows enough about the virtual environments, they know when to skip this advice and if they don't know `venv`, it's a good place to learn. Some members of our audience may not be aware that Python is written to `.py` files and executed by `python`.

jorgeorpinel replied:
Agree it's not ideal to be specific about `venv` but what's a better alternative? Just mention it and don't give out any instructions? That wouldn't help beginners. Like @iesahin mentioned advanced users will know how to replace it, probably.

iesahin replied:
If the user has visited a Github repository, IMO we can assume they know how to clone it, and for `cd`, yeah, they are very 

`venv` is Python specific, we can't assume the user knows Python.  It's true that `venv` is related to a particular workflow, cloning to your local machine and experimenting. The user could be cloning this into a container, in that case they should know to skip.

If the user is new to Python and downloaded some earlier/later version of Tensorflow to their machine, we can't be sure that the code will work as expected. 

Asking the user to work in a virtualenv is a better approach, and without `venv`, asking them to run `pip install -r requirements.txt` is dangerous, they may break some other software without knowing, and without `pip install -r requirements.txt`, how can we say them to install the requirements?

I'm assuming it's the user's second day in Python. If it's the second year, they know they can skip this.

jorgeorpinel replied:
I don't think this is very critical. But a simple option is to just add a note like in https://dvc.org/doc/install/macos#install-with-pip

iesahin commented on get-started/code-experiments/README.md:
````diff
@@ -0,0 +1,260 @@
+python -m pip install -r requirements.txt
+```
+
+`params.yaml` defines many parameters to modify with `dvc exp run --set-param/-S`
+option. For example to set the number of convolutional units in `cnn` model, you
+can use:
````
Here `cnn` refers to the value of `model.name` parameter, and incidentally `cnn` is a CNN model 😄 It may not be _the CNN model_ for the project, though. We may add additional CNN models, this one is too simple.

jorgeorpinel replied:
I think @casperdcl's suggestion was mainly on syntax and is much easier to read. Keeping `cnn` as a quote is secondary.

casperdcl replied:
updated

iesahin commented on get-started/code-experiments/README.md:
````diff
@@ -0,0 +1,260 @@
+dvc exp run -S model.cnn.conv_units=128 
+```
+
+The experiment will produce a new `metrics.json` and you can check the changes
+in metrics:
+
+```console
+dvc exp diff
+```
````
`exp diff` is used for comparing metrics and params. There is no `exp plots` or `exp metrics`. 

I may agree that's a bit confusing but `dvc exp show` shows what? A table of params and metrics.  `diff` is better than `show`, though has some strings attached from `git diff` and `diff` utility.  I'd name it `dvc exp delta`, and the other `dvc exp table`.

jorgeorpinel replied:
Maybe add a sample output to clarify how this command displays metrics.

iesahin commented on get-started/code-experiments/requirements.txt:
```diff
@@ -0,0 +1,2 @@
+tensorflow==2.4.1
```
Recent versions contain GPU support in the `tensorflow` package. 

https://www.tensorflow.org/install/gpu#pip_package

When I install from this `requirements` to a `venv`, I can see it's using `libcuda`. 

Tensorflow has some tough versioning requirements, e.g., you can't run `2.4` in Python 3.9. I changed the line to `>2.4,<2.5`, and put a version string to `ruamel`.

casperdcl commented:
It looks like this PR has already been deployed to https://github.com/iterative/get-started-experiments before being reviewed & merged :thinking:

Anyway for all these generation-type stuff we should also have a note at the bottom of generated READMEs linking back to this source repo.

iesahin commented:
I generated and pushed the repos because I think it's easier to see the end
product in review.

I see the generation code as an internal stuff while the resulting repos as
the product, but I may add the links to this repo if it feels better.

casperdcl commented:
> I generated and pushed the repos because I think it's easier to see the end product in review.

best to push to a temp repo rather than an in-production repo (otherwise defeats at least one of the objectives of review)

> I see the generation code as an internal stuff while the resulting repos as the product, but I may add the links to this repo if it feels better.

exactly. Advanced users (and us too, for our own sanity) care about internal stuff being linked (a one-line contributing section stating "please open PRs in the [source repo](https://github.com/iterative/example-repos-dev) if you find a bug," etc.)

casperdcl reviewed (commented):
- please use the "commit suggestion/add suggestion to batch" feature
- try to avoid renaming files during review (you've moved `get-started/code-experiments/` to `get-started-experiments/code-experiments/`)

not paying attention to the above means breaking the GH review linking system, making it impossible to tell what's been addressed without reviewing everything all over again.

iesahin commented:
I reverted the rename @casperdcl

jorgeorpinel commented on get-started/code-experiments/README.md:
```diff
@@ -0,0 +1,260 @@
+This repository contains a project that trains a classifier for [Fashion
+MNIST][fmnist] dataset in Tensorflow. 
+
+[fmnist]: https://github.com/zalandoresearch/fashion-mnist
```
> don't use [links][like this] unless you re-use them

On this, seems like a stylistic choice TBH (up to eahc person for now). But I would agree it's best to embed them most times unless

a) the URLs are super long
b) they are not very short and repeat (esp if more than once)
c) they contain a URL anchor e.g. `/some/path#specific-section-title`

What do you guys think?

casperdcl replied:
I don't think this repo uses "raw" markdown - it's actually "code" markdown so this makes more sense to me:

- keep `[with the link text](URL)`
- unless the URL is re-used, in which case avoid `[duplication & typos][URL]` with `[aliases][URL]`

I think we should be using language features to our advantage rather than 1. because the features exist or 2. based on style preferences

iesahin replied:
For me, the difference is this:

![levinas-210602-114952](https://user-images.githubusercontent.com/476310/120450219-ee932c80-c398-11eb-86df-1a566f3dd0e6.png)
![levinas-210602-114918](https://user-images.githubusercontent.com/476310/120450243-f521a400-c398-11eb-9d86-4326c6a95bcf.png)

I admit the [reference links] are rare in MD documents, but seeing paragraphs broken by inline links makes my OCD side suffer. :)) The broken links are a different matter; you can have broken links in (inline) links as well. 

I changed the links in iterative/dvc.org#2497 because the link checker wasn't aware of them. Here, however, we should consider people will be reading the raw `README.md` file in their editor as well. In that case having reference links is cleaner IMO.

iesahin replied:
For me, the difference is this:

![levinas-210602-114952](https://user-images.githubusercontent.com/476310/120451799-6746b880-c39a-11eb-839a-7817e1956662.png)
![levinas-210602-114918](https://user-images.githubusercontent.com/476310/120451804-6877e580-c39a-11eb-9cb0-153cf4528c08.png)


I admit the [reference links] are rare in MD documents, but seeing paragraphs broken by inline links makes my OCD side suffer. :)) The broken links are a different matter; you can have broken links in (inline) kind as well. 

I changed the links in iterative/dvc.org#2497 because the link checker wasn't aware of them. Here, however, we should consider people may be reading the raw `README.md` file in their editor as well. In that case, having reference links is cleaner IMO.

jorgeorpinel replied:
As of now I think MD link styles are a personal preference (until we manage to agree on some guidelines, lets discuss elsewhere?) so up to you, esp. in this repo.

jorgeorpinel commented on get-started/code-experiments/README.md:
````diff
@@ -0,0 +1,260 @@
+dvc exp run --queue -S model.cnn.dropout=0.9
+```
+
+and run them at once, preferably in parallel with `--jobs`:
````
"at once" may not mean what you intend. Try to use simple language when possible e.g. "run them all" here.

jorgeorpinel commented:
> > looks like this PR has already been deployed to https://github.com/iterative/get-started-experiments
>
> generated and pushed the repos because I think it's easier to see the end
product in review

Is that repo in use already? Has the actual Python project been reviewed? Did it exist prior to this PR (and if so what README did it have)? Depending on those answers it may be OK to push it but in general it does seem funny to put something in production before it's reviewed.

p.s. it's pretty easy to review here using rich diff feature:

![image](https://user-images.githubusercontent.com/1477535/120373812-6a7c6e80-c2de-11eb-8b66-10d92f7d2601.png)

> note at the bottom of generated READMEs linking back to this source repo

It's a good question. We've been fine without linking them so far and it's OK if users open issues in the generated one (we just transfer them). Rarely do people send PRs but in that case it's indeed probably annoying that we ask them to re-do the work in "right" place. That said the note wouldn't hurt so I'm for it. That said there's no guarantee it will be noticed so it may not be a solution (consider using a [PR template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) instead/additionally).

jorgeorpinel reviewed (commented):
General review on the README:

Let's add sections (**Installation**) and put the pipeline info in the intro like in https://github.com/iterative/example-get-started/blob/master/README.md? Possibly also with a figure. Also I'd consider listing the stages instead of the params and metrics (mention these in each stage). Could even have metrics and plots here.

Should we intro plain `dvc exp run` before `exp run -S`? BTW use `--set-params` the first time at least (more explicit).

Too many --queue lines 🙂 

Maybe make a Git section (**Persistent experiments**) 

**Data/Source/Model/DVC Files** -> combine into "Project structure"? Like https://github.com/iterative/example-get-started#project-structure (more visual)

iesahin commented:
> Is that repo in use already? Has the actual Python project (and scripts) been reviewed? Did it exist prior to this PR (and if so what README did it have)? Depending on those answers it may be OK to push it but in general it does seem funny to put something in production before it's reviewed.

We have a chicken-egg 🐣 situation here. This PR produces a repository that I use in a tutorial. I'm updating the project due to reviews to the tutorial, e.g., in iterative/dvc.org#2479, but if I don't have the project, I can't have persistent base to run the commands. Also I'm using the project in Katacoda, and the same updates will go there and probably break things and require a fix. 

I may use a temporary repository until the reviews are done but that will require another step to make it public. We are doing as much as possible in the public, and e.g., @aguschin 's review above wouldn't be possible if we had a private repo. 

The same is true for `get-started-checkpoints`. We have a drafted tutorial for it, I'm planning a Katacoda scenario and we have a PR here to generate the repo. These are all interdependent. I'm sure some bugs in the code will surface in Katacoda and reviews to the tutorial will require changes to the repo. 

It's true that I may do this all using small steps and tens of PRs but I'm not sure if it's the most efficient way of using our time, e.g., I'm now sure that the code in this repo runs and builds a model with a good classification accuracy. If I'd wait for reviews to this PR's code before writing the tutorial / katacoda / docker container, we'd have to wait another six months before each of these stabilize.

iesahin commented:
I need to update this PR considerably with new/renamed directories, merged files, etc. It looks we are discussing mostly the Readme file. I'll open issue(s) for the discussion points and we can continue there, is this OK? @casperdcl @jorgeorpinel

jorgeorpinel commented:
> We have a chicken-egg 🐣 situation here. This PR produces a repository that I use in a tutorial. I'm updating the project due to reviews to the tutorial, e.g., in iterative/dvc.org#2479, but if I don't have the project, I can't have persistent base to run the commands...
The same is true for get-started-checkpoints. We have a drafted tutorial for it, I'm planning a Katacoda scenario and we have a PR here to generate the repo

Probably best to have separate PRs for the code and for the README then, if that helps (and it probably will, at least for the review process). It's OK if they're not in perfect sync all the time. 🐤 

> use a temporary repository until the reviews are done

No need to overcomplicate your life that way.

> I need to update this PR considerably with new/renamed directories, merged files, etc. It looks we are discussing mostly the Readme file. I'll open issue(s) for the discussion points and we can continue there, is this OK? @casperdcl @jorgeorpinel

I'm afraid we'll lose track of existing feedback, lets please try to keep review continuity. If you wish to split everything except README into another PR that could work.

jorgeorpinel commented:
p.s. at the same time we should be reviewing high level at this point (cc @casperdcl). It's still unclear what the overall plan is for example repos anyway (I think), so this will prob change again soon. So something like this: https://github.com/iterative/example-repos-dev/pull/44#pullrequestreview-673582057 and at the end hopefully one quick syntax round.

iesahin commented:
Simplified and updated the pipeline. 

* merged `preprocess` and `evaluate` stages to `train`. We now have `prepare` ➡️ `train`. I'll keep the full pipeline only in `get-started-pipelines` for `no-cache` examples.
* `params.py` now contains a flat list of 4 parameters: instead of hierarchical `model.cnn.conv_units`, we now have `conv_units`. 
* only `acc` (CategoricalAccuracy) and `loss` set as default metrics (but others can be turned on in `params.yaml`)
* large intermediate data dependencies removed. only `raw` (31MB), `prepared` (52MB) and `model` (4MB) are cached and pushed/pulled. (There was a ~400MB `preprocessed` file.)
* merged `generate.bash` and `generate-experiments.bash` in a single file. 

Updated the repository with the new structure https://github.com/iterative/get-started-experiments/

@shcheklein @casperdcl @jorgeorpinel @dberenbaum

shcheklein commented on get-started/code-experiments/dvc.yaml:
```diff
@@ -0,0 +1,58 @@
+    - src/models.py
+    - src/train.py
+    params:
+    - model.cnn.conv_kernel_size
```
should it be regenerated? why do we have this file at all, btw?

shcheklein commented on get-started/code-experiments/params.yaml:
```diff
@@ -0,0 +1,4 @@
+epochs: 1
+conv_units: 16
+dense_units: 128
+dropout: 0.5
```
minor: newline

shcheklein commented on get-started/code-experiments/.gitignore:
```diff
@@ -0,0 +1,6 @@
+*.py[cod]
+*$py.class
+
+.venv
```
minor, new line

shcheklein commented on get-started/.gitignore:
```diff
@@ -0,0 +1,20 @@
+# Byte-compiled / optimized / DLL files
```
why do we have 2 gitignore files?

shcheklein commented on get-started/code-experiments/src/prepare.py:
```diff
@@ -0,0 +1,108 @@
+
+
+def mnist_images_idx_to_array(images_filename):
+    images_f = gzip.open(images_filename, mode="rb")
```
as we discussed - let's simplify this, let's not use `gz` and this format - let's use regular images

shcheklein commented on get-started/code-experiments/src/prepare.py:
```diff
@@ -0,0 +1,108 @@
+    return images_array
+
+
+def mnist_labels_idx_to_array(labels_filename):
```
same here - code should be simpler

shcheklein replied:
if we want to have some meaningful prepare step - I would apply some augmentations or some other image transformations

shcheklein commented on get-started/code-experiments/src/util.py:
```diff
@@ -0,0 +1,92 @@
+from ruamel.yaml import YAML
+import numpy as np
+
+FULL_PARAMS = {
```
simplicity of the code is important to my mind ... do we actually okay to keep this workaround?

shcheklein commented:
Nice @iesahin ! It's way simpler now.

> params.py now contains a flat list of 4 parameters: instead of hierarchical model.cnn.conv_units, we now have conv_units.

I would keep one nested level (to show that it's possible)

Also a few other comments:

- can we use images and labels instead of `.gz`?
- simplify the code also to some reasonable extent

shcheklein commented:
But overall, I would try at this point to write some documentation first and see what can we get there - then do other iterations on this.

iesahin commented:
I think this is done. I'm about to merge if no more changes requested. @shcheklein @jorgeorpinel @casperdcl

casperdcl commented on README.md:
```diff
@@ -39,3 +36,7 @@ $ ./deploy.sh
+
+### get-started-experiments
+
+- `generate.sh`: Generates the [repository](https://github.com/iterative/get-started-experiments) for _Get Started with Experiments_.  It creates a new project in `get-started-experiments/build/YYYY-MM-DD-HH-MM-SS/get-started-experiments` and an accompanying script to push the repository to DVC and Git. The generated repository uses `s3://dvc-public/remote/get-started-experiments/` as a DVC remot ...
```
is this the intention?

```suggestion
- `generate.sh`: Generates https://github.com/iterative/get-started-experiments (used for [_Get Started with Experiments_](dvc.org/doc/start/experiments)). It creates a new project in `get-started-experiments/build/YYYY-MM-DD-HH-MM-SS/get-started-experiments` and an accompanying script to push the repository to DVC and Git. The generated repository uses `s3://dvc-public/remote/get-started-experiments/` as a DVC remote.
```

shcheklein commented on get-started-experiments/generate.bash:
```diff
@@ -148,7 +156,7 @@ set -veux
 pushd ${REPO_PATH}
 
-dvc remote add --default storage s3://dvc-public/code/${PROJECT_NAME}/
+# dvc remote add --default storage s3://dvc-public/remote/${PROJECT_NAME}/
```
dead code?

shcheklein commented on get-started-experiments/code-experiments/.gitignore:
```diff
@@ -0,0 +1,6 @@
+*.py[cod]
+*$py.class
+
+.venv
```
minor: newline

shcheklein commented on .gitignore:
```diff
@@ -1,51 +0,0 @@
-# Byte-compiled / optimized / DLL files
```
q: why are we removing the whole file? will if affect other projects by chance?

shcheklein commented on get-started-experiments/code-experiments/README.md:
```diff
@@ -0,0 +1,142 @@
+
+### Parameters
+
+There are two parameters in the project. They are set in `params.yaml`. `models.conv_units` defines the number of convolutional units in the model, and `train.epochs` sets the number of epochs to train the model. 
```
preferable use the same style across all *.md

shcheklein reviewed (approved):
some minor comments here and there, looks good otherwise to me.

iesahin commented:
@shcheklein About the root `.gitignore` file I probably deleted the file mistakenly because I never had any intention to work on files outside of the `get-started-experiments` directory. I've just updated the README recently. I've reinstated the global `.gitignore`.
